plugin/trino-mysql/src/test/java/io/trino/plugin/mysql/TestMySqlTypeMapping.java

```diff
@@ -719,6 +622,117 @@ public void testTimestamp(ZoneId sessionZone)
+                .addRoundTrip("datetime(6)", "TIMESTAMP '1970-01-01 00:00:01.1234569'", TIMESTAMP_MILLIS, "TIMESTAMP '1970-01-01 01:00:01.123'")
+
+                .execute(getQueryRunner(), session, mysqlCreateAndInsert("tpch.test_datetime"));
+
+        // Note that MySQL treats TIMESTAMP literals as datetime type
+        SqlDataTypeTest.create()
+                // TODO support higher precision timestamps (https://github.com/trinodb/trino/issues/6910)
+                // before epoch with second fraction
+                //.addRoundTrip("timestamp(7)", "TIMESTAMP '1969-12-31 23:59:59.1230000'", createTimestampType(7), "TIMESTAMP '1969-12-31 23:59:59.1230000'")
+                //.addRoundTrip("timestamp(7)", "TIMESTAMP '1969-12-31 23:59:59.1234567'", createTimestampType(7), "TIMESTAMP '1969-12-31 23:59:59.1234567'")
+
+                // precision 0 ends up as precision 0
+                .addRoundTrip("TIMESTAMP '1970-01-01 00:00:01'", "TIMESTAMP '1970-01-01 01:00:01.000'")
+
+                .addRoundTrip("TIMESTAMP '1970-01-01 00:00:00.1'", "TIMESTAMP '1970-01-01 01:00:00.100'")
+                .addRoundTrip("TIMESTAMP '1970-01-01 00:00:00.9'", "TIMESTAMP '1970-01-01 01:00:00.900'")
+                .addRoundTrip("TIMESTAMP '1970-01-01 00:00:00.123'", "TIMESTAMP '1970-01-01 01:00:00.123'")
+                .addRoundTrip("TIMESTAMP '1970-01-01 00:00:00.123000'", "TIMESTAMP '1970-01-01 01:00:00.123'")
+                .addRoundTrip("TIMESTAMP '1970-01-01 00:00:00.999'", "TIMESTAMP '1970-01-01 01:00:00.999'")
+                // max supported precision
+                .addRoundTrip("TIMESTAMP '1970-01-01 00:00:00.123456'", "TIMESTAMP '1970-01-01 01:00:00.123'")
+
+                .addRoundTrip("TIMESTAMP '2020-09-27 12:34:56.1'", "TIMESTAMP '2020-09-27 12:34:56.100'")
+                .addRoundTrip("TIMESTAMP '2020-09-27 12:34:56.9'", "TIMESTAMP '2020-09-27 12:34:56.900'")
+                .addRoundTrip("TIMESTAMP '2020-09-27 12:34:56.123'", "TIMESTAMP '2020-09-27 12:34:56.123'")
+                .addRoundTrip("TIMESTAMP '2020-09-27 12:34:56.123000'", "TIMESTAMP '2020-09-27 12:34:56.123'")
+                .addRoundTrip("TIMESTAMP '2020-09-27 12:34:56.999'", "TIMESTAMP '2020-09-27 12:34:56.999'")
+                // max supported precision
+                .addRoundTrip("TIMESTAMP '2020-09-27 12:34:56.123456'", "TIMESTAMP '2020-09-27 12:34:56.123'")
+
+                // round down
+                .addRoundTrip("TIMESTAMP '1970-01-01 00:00:00.123451'", "TIMESTAMP '1970-01-01 01:00:00.123'")
+
+                // nanos round up, end result rounds down
+                .addRoundTrip("TIMESTAMP '1970-01-01 00:00:00.123499'", "TIMESTAMP '1970-01-01 01:00:00.123'")
+                .addRoundTrip("TIMESTAMP '1970-01-01 00:00:00.123399'", "TIMESTAMP '1970-01-01 01:00:00.123'")
+
+                // round up
+                .addRoundTrip("TIMESTAMP '1970-01-01 00:00:00.123999'", "TIMESTAMP '1970-01-01 01:00:00.124'")
```

Comment:
This would belong to `testTimestampFromMySql`.

However, with `mysqlCreateAndInsert`, i strongly recommend using explicit column type.

the implicit types are great and unambiguous when doing input via Trino

Reply:
You might not be able to do that because SqlDataType will complain that the types don't match. The inserted type is `datetime(6)` and what we read back is `TIMESTAMP(3)`. Hence cheating using literals until #6910 lands.

Reply:
> However, with mysqlCreateAndInsert, i strongly recommend using explicit column type.
> 
> the implicit types are great and unambiguous when doing input via Trino

https://github.com/trinodb/trino/pull/7245